SignalAndSlots

Ar444ri opened the pull request.

Ar444ri committed:
SignalAndSlots

hBuzzy commented on SignalsAndSlots/bar.cpp:
```diff
@@ -0,0 +1,30 @@
+#include "bar.h"
+
+Bar::Bar() {
+    const QString normalStyle =
+        "QProgressBar::chunk {background: #E6E6FA; Width: 20px; margin: 0.5px; "
+        "border: 1px solid black; border-radius:8px; Border-Radius: 4px;} "
+        "QProgressBar { text-align: center; font-size:14px; border-radius:8px; "
+        "color:black;}";
```
У вас уже есть две константы внутри `SetColor`. Зачем создавать еще одну такую же? Сделайте константы не локальными переменными, а приватными полями, тогда вы избавитесь от дубляжа кода.

Кроме того, в конструкторе можно просто вызвать `SetColor ` и метод сам решит какой цвет сейчас задать.

hBuzzy commented on SignalsAndSlots/enemy.h:
```diff
@@ -0,0 +1,23 @@
+
+ signals:
+  void MakeDamage(int gamage);
+  void MakeRestore(int gamage);
```
В параметрах лучше указать тогда вместо damage -> restore

hBuzzy commented on SignalsAndSlots/widget.cpp:
```diff
@@ -0,0 +1,91 @@
+  int restore = 5;
+
+  Player *player = new Player(maxHealth);
+  Enemy *enemy = new Enemy(demage);
+  Enemy *myenemy = new Enemy(restore);
```
Если  хотите, чтобы враг мог и наносить урон и восстанавливать (допустим), то передайте ему в конструктор оба значения. Это исключит текущую ситуацию, где у вас enemy использует одну часть функционала, а myEnemy - вторую. Вместо двух врагов можно будет создать одного.

myenemy -> myEnemy

hBuzzy commented on SignalsAndSlots/widget.cpp:
```diff
@@ -0,0 +1,91 @@
+  connect(restoreButton, &QPushButton::clicked, myenemy,&Enemy::OnRestoreButtonClicked);
+  connect(myenemy, &Enemy::MakeRestore, player, &Player::TakeRestore);
+
+  connect(myhealthBar, &QProgressBar::valueChanged, myhealthSlider,[=](int value){
+
+         myhealthBar->SetColor(value,myhealthSlider->value());
+  });
```
У вас изменение слайдера должно вызывать проверку у Progressbar, но изменение бара не должно вызывать изменение слайдера.  Данного соединения быть не должно.

hBuzzy commented on SignalsAndSlots/widget.cpp:
```diff
@@ -0,0 +1,91 @@
+      myhealthSliderCaption->setNum(value);
+  });
+
+  connect(myhealthBar, &QProgressBar::valueChanged, player, &Player::SetHealth);
```
Задание порога должно вызываться слайдером, а не баром.

hBuzzy commented on SignalsAndSlots/player.h:
```diff
@@ -0,0 +1,26 @@
+ public slots:
+  void TakeDamage(int damage);
+  void TakeRestore(int restore);
+  void SetHealth(int health);
```
За изменение здоровья отвечают методы TakeDamage и TakeRestore. SetHealth должно задавать порог (процент здоровья), а не само здоровье.

hBuzzy commented on SignalsAndSlots/widget.cpp:
```diff
@@ -0,0 +1,91 @@
+  });
+
+  connect(myhealthBar, &QProgressBar::valueChanged, player, &Player::SetHealth);
+  connect(player, &Player::HealthChanged, myhealthBar, &QProgressBar::setValue);
```
&Bar::setValue - лучше используйте свой класс вместо родителя.

hBuzzy commented on SignalsAndSlots/widget.h:
```diff
@@ -0,0 +1,25 @@
+
+ private:
+  Ui::Widget *ui;
+  Bar *myhealthBar;
+  QSlider *myhealthSlider;
```
Приватные поля именуем с нижним подчеркиванием на конце: `variableName_`. Каждое новое слово, кроме первого, начинается с большой буквы.

hBuzzy commented on SignalsAndSlots/widget.cpp:
```diff
@@ -0,0 +1,91 @@
+  mainLayout->addWidget(widget);
+  mainLayout->addItem(bottomSpacer);
+
+  connect(damageButton, &QPushButton::clicked, enemy,&Enemy::OnDamageButtonClicked);
+  connect(enemy, &Enemy::MakeDamage, player, &Player::TakeDamage);
+
+  connect(restoreButton, &QPushButton::clicked, myenemy,&Enemy::OnRestoreButtonClicked);
+  connect(myenemy, &Enemy::MakeRestore, player, &Player::TakeRestore);
```
Дубляж кода, исправьте логику с нанесением и восстановлением урона врагом.

hBuzzy reviewed (changes requested):
Исправить недочеты с логикой и стилем.

Ar444ri committed:
Add files via upload

Постаралась исправить свои ошибки, но не понимаю замечания по поводу последнего коннекта(изменение бара не должно вызывать изменение слайдера).Без этого коннекта не получается нормально контролировать изменение цвета, когда мы добавляем/убавляем HP нашего игрока. (Цвет бара остается таким же, даже если мы перешли порог здоровья и это не меняется до тех пор, пока мы снова не пошевелим слайдер, что противоречит моей логике)

Ar444ri committed:
Update bar.cpp

удалила ненужное

hBuzzy commented on SignalsAndSlots/bar.cpp:
```diff
@@ -1,30 +1,25 @@
-        "color:black;}";
-
-    this->setStyleSheet(normalStyle);
+Bar::Bar( QString danStyle, QString norStyle,QProgressBar *parent): QProgressBar(parent) {
+    danStyle_=danStyle;
+    norStyle_=norStyle;
```
Оператор присваивания, как и прочие, отделяется с двух сторон пробелами. Разделяйте параметры после запятой пробелами.

`Bar( QString danStyle` - перед `QStrting` пробел не нужен.

hBuzzy commented on SignalsAndSlots/bar.cpp:
```diff
@@ -1,30 +1,25 @@
 
-void Bar::SetColor(int health, int percent) {
-    const QString kDangerStyle =
+void Bar::SetColor(int health, int percent,const QString danStyle,const QString norStyle) {
```
Разделяйте параметры после запятой пробелами.

hBuzzy commented on SignalsAndSlots/bar.h:
```diff
@@ -8,10 +8,12 @@
+explicit Bar(QString danStyle,QString norStyle,QProgressBar *parent = nullptr);
 public slots:
- void SetColor(int health, int percent);
+ void SetColor(int health, int percent,const QString danStyle,const QString norStyle);
```
Пробелы для параметров.

`QString danStyle, QString norStyle` - не используйте непонятные и не общепринятые сокращения. Человек вне контекста задачи сомнительно, что поймет что такое `dan` и `nor`.

hBuzzy commented on SignalsAndSlots/bar.h:
```diff
@@ -8,10 +8,12 @@
- void SetColor(int health, int percent);
+ void SetColor(int health, int percent,const QString danStyle,const QString norStyle);
+  private:
+   QString danStyle_;
+   QString norStyle_;
```
`QString danStyle_, QString norStyle_` - не используйте непонятные и не общепринятые сокращения. Человек вне контекста задачи сомнительно, что поймет что такое `dan` и `nor`.

hBuzzy commented on SignalsAndSlots/god.cpp:
```diff
@@ -0,0 +1,9 @@
+
+God::God(int damage, int restore,QObject *parent) : QObject(parent) {
+  damage_ = damage;
+  restore_=restore;
```
Пробелы у оператора =

hBuzzy commented on SignalsAndSlots/god.cpp:
```diff
@@ -0,0 +1,9 @@
+#include "god.h"
+
+God::God(int damage, int restore,QObject *parent) : QObject(parent) {
```
Пробелы между параметрами

hBuzzy commented on SignalsAndSlots/god.h:
```diff
@@ -0,0 +1,24 @@
+class God : public QObject {
+  Q_OBJECT
+ public:
+  explicit God(int damage,int restore, QObject *parent = nullptr);
```
Пробел

hBuzzy commented on SignalsAndSlots/widget.cpp:
```diff
@@ -68,24 +77,23 @@ Widget::Widget(QWidget *parent) : QWidget(parent), ui(new Ui::Widget) {
-  connect(myhealthBar, &QProgressBar::valueChanged, player, &Player::SetHealth);
-  connect(player, &Player::HealthChanged, myhealthBar, &QProgressBar::setValue);
-
+ connect(healthBar, &QProgressBar::valueChanged, myhealthSlider,[=](int value){
+ healthBar->SetColor(value,myhealthSlider->value(),kDangerStyle,kNormalStyle);
+  });
```
Пробелы:

```c++
connect(healthBar, &QProgressBar::valueChanged, myhealthSlider, [=](int value) {
    healthBar->SetColor(value, myhealthSlider->value(), kDangerStyle, kNormalStyle); });
```

hBuzzy commented on SignalsAndSlots/widget.cpp:
```diff
@@ -68,24 +77,23 @@ Widget::Widget(QWidget *parent) : QWidget(parent), ui(new Ui::Widget) {
+ connect(myhealthSlider, &QSlider::valueChanged, healthBar, [=](int value) {
+      healthBar->SetColor(healthBar->value(),value,kDangerStyle,kNormalStyle);
       myhealthSliderCaption->setNum(value);
   });
```
Пробелы

hBuzzy commented on SignalsAndSlots/widget.cpp:
```diff
@@ -68,24 +77,23 @@ Widget::Widget(QWidget *parent) : QWidget(parent), ui(new Ui::Widget) {
+  connect(enemy, &God::MakeRestore, player, &Player::TakeRestore);
+  connect(player, &Player::HealthChanged, healthBar, &QProgressBar::setValue);
+  connect(  myhealthSlider, &QSlider::valueChanged,
+          [  myhealthSliderCaption](int value) {   myhealthSliderCaption->setText(QString("%1 HP ").arg(value)); }
```
Лишние пробелы.

```c++
connect(myhealthSlider
[myhealthSliderCaption]
{myhealthSliderCaption->setText(QString("%1 HP ").arg(value)); }
```

hBuzzy commented on SignalsAndSlots/widget.cpp:
```diff
@@ -68,24 +77,23 @@ Widget::Widget(QWidget *parent) : QWidget(parent), ui(new Ui::Widget) {
-  });
-  connect(myhealthSlider, &QSlider::valueChanged, myhealthBar, [=](int value) {
-      myhealthBar->SetColor(myhealthBar->value(),value);
+  connect(damageButton, &QPushButton::clicked, enemy,&God::OnDamageButtonClicked);
+  connect(enemy, &God::MakeDamage, player, &Player::TakeDamage);
+  connect(restoreButton, &QPushButton::clicked, enemy,&God::OnRestoreButtonClicked);
```
Пробел между параметрами

hBuzzy reviewed (commented):
Приму полностью первую и так (за старания), но ошибки нужно будет исправить все равно.

Почитайте лекционный материал про пробелы и общий стиль, вам это поможет в дальнейших задачах :) Кроме того, можете посмотреть материал про настройку Qt Creator, можно просто установить плагин и все отступы, переносы и т. д. будут выставляться автоматически.

Ar444ri committed:
Add files via upload

Исправила замечания

hBuzzy reviewed (commented):
Принято.
